src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs

```diff
@@ -75,42 +92,68 @@ private async Task CreateBlocksAsync()
             {
                 try
                 {
-                    // Don't mine in IBD.
-                    if (this.ibdState.IsInitialBlockDownload() && !MineDuringIBD)
+                    // Don't mine in IBD in case we are connected to any node.
+                    if (this.ibdState.IsInitialBlockDownload() && this.connectionManager.ConnectedPeers.Any())
                     {
                         int attemptDelayMs = 20_000;
                         await Task.Delay(attemptDelayMs, this.cancellation.Token).ConfigureAwait(false);
 
                         continue;
                     }
 
+                    uint timeNow = (uint) this.dateTimeProvider.GetAdjustedTimeAsUnixTimestamp();
+                    uint myTimestamp = this.slotsManager.GetMiningTimestamp(timeNow);
+
+                    if (myTimestamp < timeNow)
+                    {
+                        // Sanity check.
+                        throw new Exception("Invalid mining timestamp generated.");
+                    }
 
-                    // TODO POA check if our timestamp and if not wait for our timestamp (wait with cancellation token)
+                    uint waitingTime = myTimestamp - timeNow - 1;
 
-                    // TODO will set nonce and target to const. Check todo in PoABlockDefenition
+                    if (waitingTime > 0)
+                    {
+                        // Wait till we can mine.
+                        this.logger.LogInformation("Waiting {0} seconds before block can be mined.", waitingTime);
+                        await Task.Delay(TimeSpan.FromSeconds(waitingTime), this.cancellation.Token).ConfigureAwait(false);
+                    }
 
                     ChainedHeader tip = this.consensusManager.Tip;
 
                     BlockTemplate blockTemplate = this.blockDefinition.Build(tip);
 
+                    blockTemplate.Block.Header.Time = myTimestamp;
+
                     // Timestamp should only greater than prev one.
                     if (blockTemplate.Block.Header.Time <= tip.Header.Time)
-                        continue; // TODO POA Log
+                    {
+                        // Can happen only when target spacing had crazy low value or key was compromised and someone is mining with our key.
+                        this.logger.LogWarning("Somehow another block was connected with greater timestamp. Dropping current block.");
+                        continue;
+                    }
 
-                    // TODO sign the block with our signature
+                    // TODO POA sign the block with our signature
 
                     // Update merkle root.
                     blockTemplate.Block.UpdateMerkleRoot();
 
-                    ChainedHeader chainedHeader = await this.consensusManager.BlockMinedAsync(blockTemplate.Block).ConfigureAwait(false); // TODO POA That should also do interg vaidation
+                    // TODO POA That should also do intergity vaidation
```

Comment:
`integrity validation` but that's not that important right now.

Reply:
fixed